### PR TITLE
[Encounter] Fiery Fallout

### PR DESCRIPTION
**Pull request by ben-lear**
Fiery Fallout Encounter

https://github.com/user-attachments/assets/a3f0a816-1634-43d2-bf8f-1efac7223508

### CODE REVIEW

**Comment by flx-sta** [comment]
@ben-lear please merge the most recent `mystery-battle-events` branch for CI

File: src/data/battle-anims.ts [hunk 1]
```diff
@@ -511,6 +517,21 @@ export function initMoveAnim(scene: BattleScene, move: Moves): Promise<void> {
...
+    for (let ea = 0; ea < encounterAnimIds.length; ea++) {
+      const encounterAnimId = encounterAnimIds[ea];
+      encounterAnimFetches.push(scene.cachedFetch(`./battle-anims/encounter-${encounterAnimNames[ea].toLowerCase().replace(/\_/g, "-")}.json`)
+        .then(response => response.json())
+        .then(cas => encounterAnims.set(encounterAnimId, new AnimConfig(cas))));
+    }
+    Promise.allSettled(encounterAnimFetches).then(() => resolve());
+  });
+}
+
```
Comment (flx-sta) [inline src/data/battle-anims.ts:534]: Please try to use `async/await` as much as possible

```suggestion
export async function initEncounterAnims(scene: BattleScene): Promise<void> {
  const encounterAnimNames = Utils.getEnumKeys(EncounterAnim);
  const encounterAnimIds = Utils.getEnumValues(EncounterAnim);
  const encounterAnimFetches = [];

  for (let ea = 0; ea < encounterAnimIds.length; ea++) {
    const encounterAnimId = encounterAnimIds[ea];
    const fetchPromise = scene.cachedFetch(`./battle-anims/encounter-${encounterAnimNames[ea].toLowerCase().replace(/\_/g, "-")}.json`)
      .then(response => response.json())
      .then(cas => encounterAnims.set(encounterAnimId, new AnimConfig(cas)));
    encounterAnimFetches.push(fetchPromise);
  }

  await Promise.allSettled(encounterAnimFetches);
}

```
<sup> CODE NOT TESTED!</sup>

File: src/data/battle-anims.ts [hunk 2]
```diff
@@ -565,6 +586,12 @@ export function loadCommonAnimAssets(scene: BattleScene, startLoad?: boolean): P
   });
 }
 
+export function loadEncounterAnimAssets(scene: BattleScene, startLoad?: boolean): Promise<void> {
+  return new Promise(resolve => {
+    loadAnimAssets(scene, Array.from(encounterAnims.values()), startLoad).then(() => resolve());
+  });
+}
+
```
Comment (flx-sta) [inline src/data/battle-anims.ts:594]: Please use `async/await` where possible

```suggestion
export async function loadEncounterAnimAssets(scene: BattleScene, startLoad?: boolean): Promise<void> {
  await loadAnimAssets(scene, Array.from(encounterAnims.values()), startLoad);
}

```

Comment (flx-sta) [reply src/data/battle-anims.ts:594]: Also add some jsdoc

File: src/data/battle-anims.ts (hunk 1 above)
Comment (flx-sta) [reply src/data/battle-anims.ts:534]: Also add some jsdoc

File: src/data/battle-anims.ts [hunk 3]
```diff
@@ -102,6 +102,11 @@ export enum CommonAnim {
     LOCK_ON = 2120
 }
 
+export enum EncounterAnim {
```
Comment (flx-sta) [inline src/data/battle-anims.ts:105]: Also add some jsdoc

File: src/data/battle-anims.ts [hunk 4]
```diff
@@ -1045,6 +1251,24 @@ export class MoveChargeAnim extends MoveAnim {
   }
 }
 
+export class EncounterBattleAnim extends BattleAnim {
```
Comment (flx-sta) [inline src/data/battle-anims.ts:1254]: add jsdoc

File: src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+import { PokemonMove } from "#app/field/pokemon";
+import { Moves } from "#enums/moves";
+import { EncounterAnim, EncounterBattleAnim, initMoveAnim, loadMoveAnimAssets } from "#app/data/battle-anims";
+import { WeatherType } from "#app/data/weather";
+import { randSeedInt } from "#app/utils";
+
+/** the i18n namespace for the encounter */
+const namespace = "mysteryEncounter:fiery_fallout";
+
+export const FieryFalloutEncounter: IMysteryEncounter =
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts:22]: add jsdoc (see lost-at-sea encounter)

File: src/data/mystery-encounters/encounters/safari-zone-encounter.ts [hunk 6]
```diff
@@ -304,14 +304,14 @@ async function throwBait(scene: BattleScene, pokemon: EnemyPokemon): Promise<boo
...
-
   return new Promise(resolve => {
     scene.trainer.setTexture(`trainer_${scene.gameData.gender === PlayerGender.FEMALE ? "f" : "m"}_back_pb`);
     scene.time.delayedCall(512, () => {
+      scene.playSound("pb_throw");
+
       // Trainer throw frames
       scene.trainer.setFrame("2");
-      scene.time.delayedCall(256, () => {
+      scene.time.delayedCall(184, () => {
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/safari-zone-encounter.ts:314]: magic number? It's being reused

File: src/data/mystery-encounters/encounters/safari-zone-encounter.ts [hunk 7]
```diff
@@ -366,18 +366,18 @@ async function throwMud(scene: BattleScene, pokemon: EnemyPokemon): Promise<bool
...
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/safari-zone-encounter.ts:380]: magic number

File: src/data/mystery-encounters/mystery-encounter.ts [hunk 8]
```diff
@@ -36,6 +37,15 @@ export enum MysteryEncounterTier {
...
 }
 
+export class StartOfBattleEffect {
+  sourcePokemon?: Pokemon;
+  sourceBattlerIndex?: BattlerIndex;
+  targets: BattlerIndex[];
+  move: PokemonMove;
+  followUp?: boolean;
+  ignorePp?: boolean;
+}
```
Comment (flx-sta) [inline src/data/mystery-encounters/mystery-encounter.ts:47]: should this be an interface?

File: src/data/mystery-encounters/mystery-encounter.ts [hunk 9]
```diff
@@ -578,6 +600,16 @@ export class MysteryEncounterBuilder implements Partial<IMysteryEncounter> {
...
+  /**
+   * Can be used to perform some extra logic (usually animations) when the enemy field is finished sliding in
+   *
+   * @param onVisualsStart - synchronous callback function to perform as soon as the enemy field finishes sliding in
+   * @returns
+   */
+  withOnVisualsStart(onVisualsStart: (scene: BattleScene) => boolean): this & Required<Pick<IMysteryEncounter, "onVisualsStart">> {
+    return Object.assign(this, { onVisualsStart: onVisualsStart });
+  }
+
```
Comment (flx-sta) [inline src/data/mystery-encounters/mystery-encounter.ts:612]: I think the name could be optimized.  but I can't think of anything rn :D

File: src/overrides.ts
Comment (flx-sta) [inline src/overrides.ts:1]: Please reset your overrides!

**Review by flx-sta — CHANGES_REQUESTED** [review]

File: src/data/mystery-encounters/encounters/safari-zone-encounter.ts (hunk 6 above)
Comment (ben-lear) [reply src/data/mystery-encounters/encounters/safari-zone-encounter.ts:314]: just for specific animation timing, pretty arbitrary

File: src/data/mystery-encounters/mystery-encounter.ts (hunk 8 above)
Comment (ben-lear) [reply src/data/mystery-encounters/mystery-encounter.ts:48]: yes

File: src/data/battle-anims.ts (hunk 3 above)
Comment (ben-lear) [reply src/data/battle-anims.ts:111]: added

**Comment by ben-lear** [comment]
Created tech debt issue to complete the unit tests in a different PR, as this one is already quite large https://github.com/AsdarDevelops/PokeRogue-Events/issues/104

**Review by flx-sta — CHANGES_REQUESTED** [review]
in-between stopping the review. Finishing another time

File: public/battle-anims/encounter-magma-bg.json
[the remark this answers is not in the document]
Comment (ben-lear) [reply public/battle-anims/encounter-magma-bg.json:1]: One the one hand I agree that minifying is optimal, especially to save on payload size coming from the server. On the other hand, for now, I actually really need these to be human readable. I’m basically frankensteining existing move animations, and then tweaking/editing certain values to get them to do what I want. If these were minified, it would be far more laborious.

[the remark this answers is not in the document]
Comment (ben-lear) [reply public/battle-anims/encounter-magma-bg.json:1]: I think it’s worth bringing up a larger conversation about json minification though. I have no idea if that would impact artist workflow or not.

[the remark this answers is not in the document]
Comment (flx-sta) [reply public/battle-anims/encounter-magma-bg.json:1]: You can literally press 1 key combination to locally deminify it and then minify it back for committing. It’s stupid to let all request be large for the sole purpose of “I’ll have to read it sometimes”. That’s what you have “format file” for

File: public/images/mystery-encounters/exclaim.json
[the remark this answers is not in the document]
Comment (ben-lear) [reply public/images/mystery-encounters/exclaim.json:1]: putting a pin in these for now

File: public/battle-anims/encounter-magma-spout.json
[the remark this answers is not in the document]
Comment (ben-lear) [reply public/battle-anims/encounter-magma-spout.json:1]: putting a pin in these for now

File: public/battle-anims/encounter-magma-bg.json
[the remark this answers is not in the document]
Comment (ben-lear) [reply public/battle-anims/encounter-magma-bg.json:1]: putting a pin in these for now

File: src/battle-scene.ts [hunk 10]
```diff
@@ -2651,7 +2650,7 @@ export default class BattleScene extends SceneBase {
       return encounter;
     }
 
-    // Common / Uncommon / Rare / Super Rare
+    // Common / Great / Ultra / Rogue
     const tierWeights = [64, 40, 21, 3];
```
Comment (flx-sta) [inline src/battle-scene.ts:2654]: Couldn't you define those numbers on the `MysteryEncounterTier` enum too? Or would that cuase side effects?

```ts
export enum MysteryEncounterTier {
  COMMON=64,
  GREAT=40,
  ULTRA=21,
  ROGUE=3,
  MASTER=0, // Not currently used
}
```

File: src/data/battle-anims.ts [hunk 11]
```diff
@@ -977,13 +1025,184 @@ export abstract class BattleAnim {
...
+        [AnimFrameTarget.TARGET, new Map<AnimFrameTarget, GraphicFrameData>() ]
+      ]);
+
+      let g = 0;
+      let u = 0;
+      let t = 0;
+
+      for (const frame of frames) {
+        let x = frame.x;
+        let y = frame.y;
```
Comment (flx-sta) [inline src/data/battle-anims.ts:1042]:
```suggestion
        let { x , y } = frame;
```

File: src/data/battle-anims.ts [hunk 12]
```diff
@@ -1045,6 +1264,24 @@ export class MoveChargeAnim extends MoveAnim {
...
+    super(user, target || user);
+
+    this.encounterAnim = encounterAnim;
+  }
+
+  getAnim(): AnimConfig {
+    return encounterAnims.get(this.encounterAnim);
+  }
+
+  isOppAnim(): boolean {
```
Comment (flx-sta) [inline src/data/battle-anims.ts:1280]: Usually the word `enemy` is used. Only `overrides` uses `OPP` for some reasons

```suggestion
  isEnemyAnim(): boolean {
```

I'm also not a fan of abbreviating the word `animation` but that is a way bigger change so just thinking out loud

File: src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+
+/**
+ * Fiery Fallout encounter.
+ * @see {@link https://github.com/AsdarDevelops/PokeRogue-Events/issues/88 | GitHub Issue #88}
+ * @see For biome requirements check {@linkcode mysteryEncountersByBiome}
+ */
+export const FieryFalloutEncounter: IMysteryEncounter =
+  MysteryEncounterBuilder.withEncounterType(MysteryEncounterType.FIERY_FALLOUT)
+    .withEncounterTier(MysteryEncounterTier.COMMON)
+    .withSceneWaveRangeRequirement(40, 180) // waves 10 to 180
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts:39]:
```suggestion
    .withSceneWaveRangeRequirement(40, 180)
```

that comment feels a bit redundant

File: src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+        for (const pkm of nonFireTypes) {
+          const percentage = DAMAGE_PERCENTAGE / 100;
+          const damage = Math.floor(pkm.getMaxHp() * percentage);
+          applyDamageToPokemon(scene, pkm, damage);
+        }
+
+        // Burn random member
+        const burnable = nonFireTypes.filter(p => isNullOrUndefined(p.status) || isNullOrUndefined(p.status.effect) || p.status?.effect === StatusEffect.BURN);
+        if (burnable?.length > 0) {
+          const chosenPokemon = burnable[randSeedInt(burnable.length - 1)];
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts:189]:
```suggestion
          const roll = randSeedInt(burnable.length);
          const chosenPokemon = burnable[roll];
```
Please separate those.
I'm also not sure if you should do `-1` since `randSeedInt(1)` rolls `0 - 0` and `randSeedInt(2)` rolls `0 - 1` (if I'm not mistaken

File: src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+
+        // No rewards
+        leaveEncounterWithoutBattle(scene, true);
+      }
+    )
+    .withOption(
+      new MysteryEncounterOptionBuilder()
+        .withOptionMode(EncounterOptionMode.DISABLED_OR_SPECIAL)
+        .withPrimaryPokemonRequirement(new TypeRequirement(Type.FIRE, true,1)) // Will set option3PrimaryName dialogue token automatically
+        .withSecondaryPokemonRequirement(new TypeRequirement(Type.FIRE, true,1)) // Will set option3SecondaryName dialogue token automatically
```
Comment (flx-sta) [inline src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts:205]:
```suggestion
        .withPrimaryPokemonRequirement(new TypeRequirement(Type.FIRE, true, 1)) // Will set option3PrimaryName dialogue token automatically
        .withSecondaryPokemonRequirement(new TypeRequirement(Type.FIRE, true, 1)) // Will set option3SecondaryName dialogue token automatically
```

File: src/data/mystery-encounters/mystery-encounter.ts [hunk 16]
```diff
@@ -18,14 +18,16 @@ import {
...
+import { EncounterAnim } from "#app/data/battle-anims";
 
 export enum MysteryEncounterVariant {
   DEFAULT,
   TRAINER_BATTLE,
   WILD_BATTLE,
   BOSS_BATTLE,
   NO_BATTLE,
-  SAFARI_BATTLE
+  REPEATED_ENCOUNTER
```
Comment (flx-sta) [inline src/data/mystery-encounters/mystery-encounter.ts:30]: not an optimal name.
How about `MULTIPLE_PHASES` for now?

File: src/data/mystery-encounters/mystery-encounter.ts [hunk 17]
```diff
@@ -90,6 +111,11 @@ export default interface IMysteryEncounter {
    * You probably shouldn't do anything with this unless you have a very specific need
    */
   introVisuals?: MysteryEncounterIntroVisuals;
+  /**
+   * Used for keeping RNG consistent on session resets, but increments when cycling through multiple "Encounters" on the same wave
+   * You should never need to modify this
+   */
+  seedOffset?: any;
```
Comment (flx-sta) [inline src/data/mystery-encounters/mystery-encounter.ts:118]: If you never have to modify this, is it possible to make it `readonly`?

File: src/data/mystery-encounters/mystery-encounter.ts
Comment (flx-sta) [inline src/data/mystery-encounters/mystery-encounter.ts:1]: I've added some `with...` method back then without providing a return type. Could you check those too? We can also do that in a separate PR if you want

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts [hunk 18]
```diff
@@ -26,29 +26,59 @@ import PokemonSpecies from "../../pokemon-species";
 import { Status, StatusEffect } from "../../status-effect";
 import { TrainerConfig, trainerConfigs, TrainerSlot } from "../../trainer-config";
 import { MysteryEncounterVariant } from "../mystery-encounter";
+import { Gender } from "#app/data/gender";
+import { Moves } from "#enums/moves";
+import { initMoveAnim, loadMoveAnimAssets } from "#app/data/battle-anims";
+
+export function doTrainerExclamation(scene: BattleScene) {
```
Comment (flx-sta) [inline src/data/mystery-encounters/utils/encounter-phase-utils.ts:33]: jsdoc

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts [hunk 19]
```diff
@@ -411,7 +476,7 @@ export function setEncounterExp(scene: BattleScene, participantId: integer | int
     const nonFaintedPartyMembers = party.filter(p => p.hp);
     const expPartyMembers = nonFaintedPartyMembers.filter(p => p.level < scene.getMaxExpLevel());
     const partyMemberExp = [];
-    let expValue = baseExpValue * (useWaveIndex ? scene.currentBattle.waveIndex : 1);
+    let expValue = Math.floor(baseExpValue * (useWaveIndex ? scene.currentBattle.waveIndex : 1) / 5 + 1);
```
Comment (flx-sta) [inline src/data/mystery-encounters/utils/encounter-phase-utils.ts:479]: where are all these numbers coming from?
Could you put up some explanation?

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts [hunk 20]
```diff
@@ -547,23 +614,40 @@ export function handleMysteryEncounterVictory(scene: BattleScene, addHealPhase:
   }
 }
 
-export function hideMysteryEncounterIntroVisuals(scene: BattleScene): Promise<boolean> {
+/**
+ *
```
Comment (flx-sta) [inline src/data/mystery-encounters/utils/encounter-phase-utils.ts:618]: text?

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts [hunk 21]
```diff
@@ -547,23 +614,40 @@ export function handleMysteryEncounterVictory(scene: BattleScene, addHealPhase:
...
+      }
+
+      // Transition
       scene.tweens.add({
         targets: introVisuals,
-        x: "+=16",
-        y: "-=16",
-        alpha: 0,
+        x: hide ? "+=16" : "-=16",
+        y: hide ? "-=16" : "+=16",
```
Comment (flx-sta) [inline src/data/mystery-encounters/utils/encounter-phase-utils.ts:640]:
```suggestion
        x: `${hide? "+" : "-"}=16`,
        y: `${hide ? "-" : "+"}=16`,
```
is that better ?

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts [hunk 22]
```diff
@@ -547,23 +614,40 @@ export function handleMysteryEncounterVictory(scene: BattleScene, addHealPhase:
...
+        alpha: hide ? 0 : 1,
         ease: "Sine.easeInOut",
-        duration: 750,
+        duration: duration,
```
Comment (flx-sta) [inline src/data/mystery-encounters/utils/encounter-phase-utils.ts:643]:
```suggestion
        duration,
```

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts [hunk 23]
```diff
@@ -573,6 +657,39 @@ export function hideMysteryEncounterIntroVisuals(scene: BattleScene): Promise<bo
   });
 }
 
+export function handleEncounterStartOfBattleEffects(scene: BattleScene) {
```
Comment (flx-sta) [inline src/data/mystery-encounters/utils/encounter-phase-utils.ts:660]: jsdoc

File: src/field/mystery-encounter-intro.ts [hunk 24]
```diff
@@ -345,6 +348,13 @@ export default class MysteryEncounterIntroVisuals extends Phaser.GameObjects.Con
       this.untint(tintSprite, duration, ease);
     });
   }
+
+  setVisible(value: boolean): this {
```
Comment (flx-sta) [inline src/field/mystery-encounter-intro.ts:352]: public method => jsdoc

File: src/field/pokemon.ts [hunk 25]
```diff
@@ -101,6 +101,8 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
   public battleSummonData: PokemonBattleSummonData;
   public turnData: PokemonTurnData;
 
+  public mysteryEncounterBattleEffects: (pokemon: Pokemon) => void = null;
```
Comment (flx-sta) [inline src/field/pokemon.ts:104]: jsdoc

File: src/phases/mystery-encounter-phases.ts [hunk 26]
```diff
@@ -108,9 +109,9 @@ export class MysteryEncounterPhase extends Phase {
         }
 
         if (title) {
-          this.scene.ui.showDialogue(text, title, null, nextAction, 0, i === 0 ? 750 : 0);
+          this.scene.ui.showDialogue(text, title, null, nextAction, 0, i === 0 ? 300 : 0);
         } else {
-          this.scene.ui.showText(text, null, nextAction, i === 0 ? 750 : 0, true);
+          this.scene.ui.showText(text, null, nextAction, i === 0 ? 300 : 0, true);
```
Comment (flx-sta) [inline src/phases/mystery-encounter-phases.ts:114]: 350 seems to be a `reused` number. Magic number?

File: src/phases/mystery-encounter-phases.ts [hunk 27]
```diff
@@ -165,6 +166,28 @@ export class MysteryEncounterOptionSelectedPhase extends Phase {
...
+ * Runs at the beginning of an Encounter's battle
+ * Will cleanup any residual flinches, Endure, etc. that are left over from startOfBattleEffects
+ * See [TurnEndPhase](../phases.ts) for more details
+ */
+export class MysteryEncounterBattleStartCleanupPhase extends Phase {
+  constructor(scene: BattleScene) {
+    super(scene);
+  }
+
+  start() {
```
Comment (flx-sta) [inline src/phases/mystery-encounter-phases.ts:179]: public method => jsdoc

File: src/phases/mystery-encounter-phases.ts [hunk 28]
```diff
@@ -240,7 +266,7 @@ export class MysteryEncounterBattlePhase extends Phase {
           this.endBattleSetup(scene);
         };
         if (!scene.currentBattle.mysteryEncounter.hideBattleIntroMessage) {
-          scene.ui.showText(this.getBattleMessage(scene), null, doTrainerSummon, 1500, true);
+          scene.ui.showText(this.getBattleMessage(scene), null, doTrainerSummon, 1000, true);
```
Comment (flx-sta) [inline src/phases/mystery-encounter-phases.ts:269]: what is the `1000`?

File: src/test/mystery-encounter/encounterTestUtils.ts [hunk 29]
```diff
@@ -1,17 +1,20 @@
...
-import { MysteryEncounterPhase, MysteryEncounterRewardsPhase } from "#app/phases/mystery-encounter-phase";
+import { CommandPhase, MessagePhase, VictoryPhase } from "#app/phases";
+import { MysteryEncounterPhase, MysteryEncounterRewardsPhase } from "#app/phases/mystery-encounter-phases";
 import MysteryEncounterUiHandler from "#app/ui/mystery-encounter-ui-handler";
 import { Mode } from "#app/ui/ui";
 import GameManager from "../utils/gameManager";
+import MessageUiHandler from "#app/ui/message-ui-handler";
+import { Status, StatusEffect } from "#app/data/status-effect";
+
+export async function runSelectMysteryEncounterOption(game: GameManager, optionNo: number, isBattle: boolean = false) {
```
Comment (flx-sta) [inline src/test/mystery-encounter/encounterTestUtils.ts:10]: jsdoc

File: src/test/mystery-encounter/encounterTestUtils.ts [hunk 30]
```diff
@@ -20,35 +23,73 @@ export async function runSelectMysteryEncounterOption(game: GameManager, optionN
...
+  });
+
+  if (isBattle) {
+    await game.phaseInterceptor.to(CommandPhase);
+  } else {
+    await game.phaseInterceptor.to(MysteryEncounterRewardsPhase);
+  }
+}
+
+export async function skipBattleRunMysteryEncounterRewardsPhase(game: GameManager) {
```
Comment (flx-sta) [inline src/test/mystery-encounter/encounterTestUtils.ts:84]: jsdoc

File: src/test/utils/TextInterceptor.ts [hunk 31]
```diff
@@ -11,6 +11,11 @@ export default class TextInterceptor {
     this.logs.push(text);
   }
 
+  showDialogue(text: string, name: string, delay?: integer, callback?: Function, callbackDelay?: integer, promptDelay?: integer): void {
```
Comment (flx-sta) [inline src/test/utils/TextInterceptor.ts:14]: jsdoc

File: src/test/utils/overridesHelper.ts [hunk 32]
```diff
@@ -1,61 +1,101 @@
...
     this.log(`Mystery encounter chance set to ${percentage}% (=${rate})!`);
+    return spy;
+  }
+
+  /**
+   * Override the encounter that spawns for the scene
+   * @param encounterType
+   * @returns spy instance
+   */
+  mysteryEncounter(encounterType: MysteryEncounterType): MockInstance {
```
Comment (flx-sta) [inline src/test/utils/overridesHelper.ts:36]:
```suggestion
  mysteryEncounter(encounterType: MysteryEncounterType) {
```
let it self-inject the type. That is a more specific version of MockInstance

File: src/test/utils/overridesHelper.ts [hunk 33]
```diff
@@ -1,61 +1,101 @@
...
+    this.game = game;
+  }
 
   /**
    * Override the encounter chance for a mystery encounter.
    * @param percentage the encounter chance in %
+   * @returns spy instance
    */
-  mysteryEncounterChance(percentage: number) {
+  mysteryEncounterChance(percentage: number): MockInstance {
```
Comment (flx-sta) [inline src/test/utils/overridesHelper.ts:23]:
```suggestion
  mysteryEncounterChance(percentage: number) {
```

let it self-inject the type. That is a more specific version of MockInstance

File: src/test/utils/overridesHelper.ts [hunk 34]
```diff
@@ -1,61 +1,101 @@
...
   }
 
   /**
    * Override the starting wave (index)
    * @param wave the wave (index) to set. Classic: `1`-`200`
+   * @returns spy instance
    */
-  startingWave(wave: number) {
-    vi.spyOn(Overrides, "STARTING_WAVE_OVERRIDE", "get").mockReturnValue(wave);
+  startingWave(wave: number): MockInstance {
```
Comment (flx-sta) [inline src/test/utils/overridesHelper.ts:57]:
```suggestion
  startingWave(wave: number) {
```

let it self-inject the type. That is a more specific version of MockInstance

File: src/test/utils/overridesHelper.ts [hunk 35]
```diff
@@ -1,61 +1,101 @@
...
+  /**
+   * Override each wave to have or not have standard trainer battles
+   * @returns spy instance
+   * @param isTrainer
+   */
+  trainerWave(isTrainer: boolean): MockInstance {
+    const spy = vi.spyOn(this.game.scene.gameMode, "isWaveTrainer").mockReturnValue(isTrainer);
+    this.log(`${isTrainer? "forcing" : "ignoring"} trainer waves!`);
+    return spy;
   }
```
Comment (flx-sta) [inline src/test/utils/overridesHelper.ts:72]:
```suggestion
  disableTrainerWaves(disable: boolean) {
    const spy = vi.spyOn(this.game.scene.gameMode, "isWaveTrainer").mockReturnValue(!disable);
    this.log(`Trainer waves are ${disabled? "disabled": "enabled"}!`);
    return spy;
  }
```

File: src/test/utils/overridesHelper.ts [hunk 36]
```diff
@@ -1,61 +1,101 @@
...
 
   /**
    * Override the weather (type)
    * @param type weather type to set
+   * @returns spy instance
    */
-  weather(type: WeatherType) {
-    vi.spyOn(Overrides, "WEATHER_OVERRIDE", "get").mockReturnValue(type);
+  weather(type: WeatherType): MockInstance {
```
Comment (flx-sta) [inline src/test/utils/overridesHelper.ts:79]:
```suggestion
  weather(type: WeatherType) {
```


let it self-inject the type. That is a more specific version of MockInstance

File: src/test/utils/overridesHelper.ts [hunk 37]
```diff
@@ -1,61 +1,101 @@
...
 
   /**
    * Override the seed
-   * @warning The seed will not be overridden unless you call `workaround_reInitSceneWithOverrides()` (testUtils)
    * @param seed the seed to set
+   * @returns spy instance
    */
-  seed(seed: string) {
-    vi.spyOn(Overrides, "SEED_OVERRIDE", "get").mockReturnValue(seed);
+  seed(seed: string): MockInstance {
```
Comment (flx-sta) [inline src/test/utils/overridesHelper.ts:90]:
```suggestion
  seed(seed: string) {
```


let it self-inject the type. That is a more specific version of MockInstance

**Review by flx-sta — CHANGES_REQUESTED** [review]
90% NIT since we are aiming for high quality code (sorry)

File: src/battle-scene.ts (hunk 10 above)
Comment (ben-lear) [reply src/battle-scene.ts:2654]: [These numbers are decremented over time throughout a session](https://github.com/AsdarDevelops/PokeRogue-Events/pull/97/files/670013a1076562698db6220777577d65772d0c4f#diff-4b1a5f9160d1d8aecce0db85774f20e3c254f97c27cc0d0473911d5cbd565396R2657-R2664), so wouldn't really work well I think.

Comment (ben-lear) [reply src/battle-scene.ts:2654]: not a bad idea

File: src/data/battle-anims.ts (hunk 12 above)
Comment (ben-lear) [reply src/data/battle-anims.ts:1280]: unfortunately, this is bad naming leftover from base class `BattleAnim`. Would have to refactor all animation classes to rename

File: src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts (hunk 14 above)
Comment (ben-lear) [reply src/data/mystery-encounters/encounters/fiery-fallout-encounter.ts:189]: you're right, the length - 1 was a brainfart.

File: src/data/battle-anims.ts (hunk 12 above)
Comment (flx-sta) [reply src/data/battle-anims.ts:1279]: don't do it then

File: src/data/mystery-encounters/mystery-encounter.ts
Comment (ben-lear) [reply src/data/mystery-encounters/mystery-encounter.ts:1]: I checked on those, TS was yelling at me for lack of return type

File: src/data/mystery-encounters/mystery-encounter.ts (hunk 16 above)
Comment (ben-lear) [reply src/data/mystery-encounters/mystery-encounter.ts:31]: maybe `CONTINUOUS_ENCOUNTER` instead?

File: src/data/mystery-encounters/mystery-encounter.ts (hunk 17 above)
Comment (ben-lear) [reply src/data/mystery-encounters/mystery-encounter.ts:122]: ended up moving to the class and making private, since it's necessary to increment

File: src/data/mystery-encounters/utils/encounter-phase-utils.ts (hunk 20 above)
Comment (ben-lear) [reply src/data/mystery-encounters/utils/encounter-phase-utils.ts:618]: tweens doesn't interpret numbers properly for some reason

File: src/phases/mystery-encounter-phases.ts (hunk 27 above)
Comment (ben-lear) [reply src/phases/mystery-encounter-phases.ts:179]: jsdoc is on the class, start is consistent to all phases

File: src/phases/mystery-encounter-phases.ts (hunk 28 above)
Comment (ben-lear) [reply src/phases/mystery-encounter-phases.ts:269]: delay before callback from the dialogue

File: src/phases/mystery-encounter-phases.ts (hunk 26 above)
Comment (ben-lear) [reply src/phases/mystery-encounter-phases.ts:114]: fixed

File: src/test/utils/TextInterceptor.ts (hunk 31 above)
Comment (ben-lear) [reply src/test/utils/TextInterceptor.ts:14]: added doc to class instead

**Review by flx-sta — APPROVED** [review]
Lgtm (not play tested)
